// pacman-react/dist/Board/Walls/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }();

exports.default = Walls;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _constants = require('../../constants');

var _Line = require('../../grid/Line');

var _Line2 = _interopRequireDefault(_Line);

var _walls = require('./walls.json');

var _walls2 = _interopRequireDefault(_walls);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function Walls(props) {
    var lineProps = {
        strokeWidth: 1,
        stroke: _constants.WALL_COLOUR,
        fill: 'none'
    };

    var linesWalls = Object.keys(_walls2.default).map(function (key) {
        var parts = _walls2.default[key].parts.map(function (_ref) {
            var _ref2 = _slicedToArray(_ref, 3),
                distance = _ref2[0],
                direction = _ref2[1],
                radius = _ref2[2];

            return { distance: distance, direction: direction, radius: radius };
        });

        return _react2.default.createElement(_Line2.default, _extends({ key: key }, props, lineProps, {
            start: _walls2.default[key].start, parts: parts }));
    });

    return _react2.default.createElement(
        'g',
        { className: 'walls' },
        linesWalls
    );
}
